agent/agent.go

```diff
@@ -563,12 +574,59 @@ func (r *agent) updateStatus(ctx context.Context) error {
 	if status == nil {
 		return nil
 	}
+
 	if err := r.Cache.UpdateStatus(status); err != nil {
 		return trace.Wrap(err, "error updating system status in cache")
 	}
+
 	return nil
 }
 
+// recordConnectionStatusChanges records any connection status changes to the
+// cluster timeline.
+func (r *agent) recordConnectionStatusChanges(ctx context.Context) error {
+	if !hasRoleMaster(r.Tags) {
+		return nil
+	}
+
+	members, err := r.ClusterMembership.Members()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	changes := diffConnectionStatuses(r.Clock, r.connectionStatuses, members)
+	if err := r.ClusterTimeline.RecordEvents(ctx, changes); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// diffConnectionStatuses returns the differences of the connection statuses
+// stored in prevStatuses and the statuses of the currMembers. The diff is
+// returned as a list of timeline events.
+func diffConnectionStatuses(clock clockwork.Clock, prevStatuses map[string]string,
+	currMembers []membership.ClusterMember) (events []*pb.TimelineEvent) {
+	for _, member := range currMembers {
+		prevStatus, exists := prevStatuses[member.Name()]
+		prevStatuses[member.Name()] = member.Status()
+
+		// If the status has not changed or this is a new member, no need to record an event.
```

Comment:
Why not record an "online" event for a new member (e.g. when a new node joins) as well? If I understood this comment correctly.